File: core/internal/workflow/status.go
// Package workflow provides plan workflow status utilities.
package workflow

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
)

// PhaseStatus categorizes a single phase.
type PhaseStatus struct {
	Phase     int    `json:"phase"`
	File      string `json:"file"`
	State     string `json:"state"` // completed, in-progress, pending
	TodoTotal int    `json:"todoTotal"`
	TodoDone  int    `json:"todoDone"`
}

// PhaseCounts holds aggregate phase counts by state.
type PhaseCounts struct {
	Total      int `json:"total"`
	Completed  int `json:"completed"`
	InProgress int `json:"inProgress"`
	Pending    int `json:"pending"`
}

// WorkflowStatus is the JSON output of GetStatus.
type WorkflowStatus struct {
	PlanDir  string       `json:"planDir"`
	Status   string       `json:"status"` // completed, in-progress, pending
	Progress int          `json:"progress"` // 0-100
	Phases   PhaseCounts  `json:"phases"`
	Reports  int          `json:"reports"`
	Detail   []PhaseStatus `json:"detail,omitempty"`
}

// GetStatus returns the workflow status for the given plan directory.
func GetStatus(planDir string) (*WorkflowStatus, error) {
	entries, err := os.ReadDir(planDir)
	if err != nil {
		return nil, fmt.Errorf("read plan dir: %w", err)
	}

	result := &WorkflowStatus{PlanDir: planDir}

	for _, e := range entries {
		if e.IsDir() || !strings.HasPrefix(e.Name(), "phase-") || !strings.HasSuffix(e.Name(), ".md") {
			continue
		}
		ps, err := categorizePhase(planDir, e.Name())
		if err != nil {
			return nil, fmt.Errorf("categorize %s: %w", e.Name(), err)
		}
		result.Detail = append(result.Detail, ps)
		result.Phases.Total++
		switch ps.State {
		case "completed":
			result.Phases.Completed++
		case "in-progress":
			result.Phases.InProgress++
		default:
			result.Phases.Pending++
		}
	}

	// Calculate progress percentage
	if result.Phases.Total > 0 {
		result.Progress = (result.Phases.Completed * 100) / result.Phases.Total
	}

	// Overall status
	switch {
	case result.Phases.Total == 0:
		result.Status = "pending"
	case result.Phases.Completed == result.Phases.Total:
		result.Status = "completed"
	case result.Phases.Completed > 0 || result.Phases.InProgress > 0:
		result.Status = "in-progress"
	default:
		result.Status = "pending"
	}

	// Count reports
	result.Reports = countReportFiles(planDir)

	return result, nil
}

// categorizePhase reads a phase file and determines its completion state.
func categorizePhase(planDir, fname string) (PhaseStatus, error) {
	data, err := os.ReadFile(filepath.Join(planDir, fname))
	if err != nil {
		return PhaseStatus{}, err
	}

	ps := PhaseStatus{
		Phase: extractPhaseNumFromName(fname),
		File:  fname,
	}

	for _, line := range strings.Split(string(data), "\n") {
		trimmed := strings.TrimSpace(line)
		if strings.HasPrefix(trimmed, "- [x]") || strings.HasPrefix(trimmed, "- [X]") {
			ps.TodoDone++
			ps.TodoTotal++
		} else if strings.HasPrefix(trimmed, "- [ ]") {
			ps.TodoTotal++
		}
	}

	switch {
	case ps.TodoTotal == 0:
		ps.State = "pending"
	case ps.TodoDone == ps.TodoTotal:
		ps.State = "completed"
	case ps.TodoDone > 0:
		ps.State = "in-progress"
	default:
		ps.State = "pending"
	}

	return ps, nil
}

// extractPhaseNumFromName parses the phase number from a filename like phase-03-xxx.md.
func extractPhaseNumFromName(fname string) int {
	parts := strings.SplitN(strings.TrimPrefix(fname, "phase-"), "-", 2)
	if len(parts) == 0 {
		return 0
	}
	n := 0
	for _, ch := range parts[0] {
		if ch < '0' || ch > '9' {
			break
		}
		n = n*10 + int(ch-'0')
	}
	return n
}

// countReportFiles counts files in reports/ and research/ subdirectories.
func countReportFiles(planDir string) int {
	count := 0
	for _, sub := range []string{"reports", "research"} {
		entries, err := os.ReadDir(filepath.Join(planDir, sub))
		if err != nil {
			continue
		}
		for _, e := range entries {
			if !e.IsDir() {
				count++
			}
		}
	}
	return count
}

File: crates/solon-core/src/workflow/status.rs
/// Plan workflow status: phase completion tracking and progress calculation.
/// Ported from Go solon-core/internal/workflow/status.go
use anyhow::{anyhow, Result};
use serde::{Deserialize, Serialize};
use std::path::Path;

/// Status of a single phase.
#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PhaseStatus {
    pub phase: usize,
    pub file: String,
    /// "completed", "in-progress", or "pending"
    pub state: String,
    pub todo_total: usize,
    pub todo_done: usize,
}

/// Aggregate phase counts by state.
#[derive(Debug, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PhaseCounts {
    pub total: usize,
    pub completed: usize,
    pub in_progress: usize,
    pub pending: usize,
}

/// JSON output of GetStatus.
#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct WorkflowStatus {
    pub plan_dir: String,
    /// "completed", "in-progress", or "pending"
    pub status: String,
    /// 0-100
    pub progress: usize,
    pub phases: PhaseCounts,
    pub reports: usize,
    pub detail: Vec<PhaseStatus>,
}

/// Return workflow status for the given plan directory.
pub fn get_status(plan_dir: &str) -> Result<WorkflowStatus> {
    let dir = Path::new(plan_dir);
    let entries = std::fs::read_dir(dir).map_err(|e| anyhow!("read plan dir: {}", e))?;

    let mut result = WorkflowStatus {
        plan_dir: plan_dir.to_string(),
        status: String::new(),
        progress: 0,
        phases: PhaseCounts::default(),
        reports: 0,
        detail: Vec::new(),
    };

    let mut files: Vec<String> = entries
        .flatten()
        .filter(|e| {
            let name = e.file_name().to_string_lossy().to_string();
            !e.path().is_dir() && name.starts_with("phase-") && name.ends_with(".md")
        })
        .map(|e| e.file_name().to_string_lossy().to_string())
        .collect();
    files.sort();

    for fname in files {
        let ps = categorize_phase(plan_dir, &fname)
            .map_err(|e| anyhow!("categorize {}: {}", fname, e))?;
        result.phases.total += 1;
        match ps.state.as_str() {
            "completed" => result.phases.completed += 1,
            "in-progress" => result.phases.in_progress += 1,
            _ => result.phases.pending += 1,
        }
        result.detail.push(ps);
    }

    // Calculate progress percentage
    if result.phases.total > 0 {
        result.progress = (result.phases.completed * 100) / result.phases.total;
    }

    // Overall status
    result.status = if result.phases.total == 0 {
        "pending".to_string()
    } else if result.phases.completed == result.phases.total {
        "completed".to_string()
    } else if result.phases.completed > 0 || result.phases.in_progress > 0 {
        "in-progress".to_string()
    } else {
        "pending".to_string()
    };

    result.reports = count_report_files(plan_dir);
    Ok(result)
}

fn categorize_phase(plan_dir: &str, fname: &str) -> Result<PhaseStatus> {
    let content = std::fs::read_to_string(Path::new(plan_dir).join(fname))?;

    let mut ps = PhaseStatus {
        phase: extract_phase_num(fname),
        file: fname.to_string(),
        state: String::new(),
        todo_total: 0,
        todo_done: 0,
    };

    for line in content.lines() {
        let t = line.trim();
        if t.starts_with("- [x]") || t.starts_with("- [X]") {
            ps.todo_done += 1;
            ps.todo_total += 1;
        } else if t.starts_with("- [ ]") {
            ps.todo_total += 1;
        }
    }

    ps.state = if ps.todo_total == 0 {
        "pending".to_string()
    } else if ps.todo_done == ps.todo_total {
        "completed".to_string()
    } else if ps.todo_done > 0 {
        "in-progress".to_string()
    } else {
        "pending".to_string()
    };

    Ok(ps)
}

fn extract_phase_num(fname: &str) -> usize {
    let stripped = fname.strip_prefix("phase-").unwrap_or(fname);
    stripped
        .chars()
        .take_while(|c| c.is_ascii_digit())
        .collect::<String>()
        .parse()
        .unwrap_or(0)
}

fn count_report_files(plan_dir: &str) -> usize {
    let mut count = 0;
    for sub in &["reports", "research"] {
        if let Ok(entries) = std::fs::read_dir(Path::new(plan_dir).join(sub)) {
            count += entries.flatten().filter(|e| !e.path().is_dir()).count();
        }
    }
    count
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_extract_phase_num() {
        assert_eq!(extract_phase_num("phase-03-research.md"), 3);
        assert_eq!(extract_phase_num("phase-10-final.md"), 10);
    }

    #[test]
    fn test_get_status_all_pending() {
        let tmp = tempfile::tempdir().unwrap();
        let dir = tmp.path().to_str().unwrap();
        std::fs::write(tmp.path().join("phase-01-a.md"), "## TODO\n\n- [ ] Task\n").unwrap();
        std::fs::write(tmp.path().join("phase-02-b.md"), "## TODO\n\n- [ ] Task\n").unwrap();

        let s = get_status(dir).unwrap();
        assert_eq!(s.status, "pending");
        assert_eq!(s.progress, 0);
        assert_eq!(s.phases.total, 2);
        assert_eq!(s.phases.pending, 2);
    }

    #[test]
    fn test_get_status_completed() {
        let tmp = tempfile::tempdir().unwrap();
        let dir = tmp.path().to_str().unwrap();
        std::fs::write(tmp.path().join("phase-01-a.md"), "- [x] Done\n").unwrap();

        let s = get_status(dir).unwrap();
        assert_eq!(s.status, "completed");
        assert_eq!(s.progress, 100);
    }

    #[test]
    fn test_get_status_in_progress() {
        let tmp = tempfile::tempdir().unwrap();
        let dir = tmp.path().to_str().unwrap();
        std::fs::write(tmp.path().join("phase-01-a.md"), "- [x] Done\n").unwrap();
        std::fs::write(tmp.path().join("phase-02-b.md"), "- [ ] Pending\n").unwrap();

        let s = get_status(dir).unwrap();
        assert_eq!(s.status, "in-progress");
        assert_eq!(s.progress, 50);
    }

    #[test]
    fn test_count_report_files() {
        let tmp = tempfile::tempdir().unwrap();
        let dir = tmp.path().to_str().unwrap();
        std::fs::create_dir(tmp.path().join("reports")).unwrap();
        std::fs::write(tmp.path().join("reports").join("report-1.md"), "").unwrap();
        std::fs::write(tmp.path().join("reports").join("report-2.md"), "").unwrap();
        assert_eq!(count_report_files(dir), 2);
    }
}
